Projects/Shapes/Shapes/Shapes/ShapeBuilder.cpp

```diff
@@ -0,0 +1,369 @@
+	std::cout << "1 - BuiltIn Array" << std::endl;
+	std::cout << "2 - Static Array" << std::endl;
+	std::cout << "3 - Dynamic Array" << std::endl;
+	std::cout << "4 - Exit" << std::endl;
+
+	int Option = GetValidIntInput(1, 4);
+
+	ActiveCommand = (EMenuCommand)Option;
+
+	std::cin.clear();
+	std::cin.ignore(INT_MAX, '\n');
+}
+
+
+//-------------------------------------------------------------------------------------------
+//										BUILT IN
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::AskForShapesBuiltIn()
+{
+	system("CLS");
+
+	std::cout << "We're gonna add your shapes to a built in array!" << std::endl;
+	std::cout << "How many shapes would you like to add?" << std::endl;
+
+
+	//Fill Array
+	int NumberOfShapes = GetValidIntInput(0, INT_MAX);
+	FShape** Shapes = FillBuiltInArray(NumberOfShapes);
+
+	//Print Array
+	PrintArray(Shapes, NumberOfShapes);
+	RequestEnterFromInput();
+
+	//Delete
+	DeleteArray(Shapes, NumberOfShapes);
+	
+	ActiveCommand = EMenuCommand::MainMenu;
+	
+}
+
+//-------------------------------------------------------------------------------------------
+FShape** FShapeBuilder::FillBuiltInArray(const int NumberOfShapes) const
+{
+	FShape** Shapes = new FShape * [NumberOfShapes];
+
+	for (int i = 0; i < NumberOfShapes; i++)
+	{
+		system("CLS");
+		Shapes[i] = CreateShape();
+	}
+
+	return Shapes;
+}
+
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::PrintArray(FShape** Array, const int NumberOfShapes) const
+{
+	system("CLS");
+
+	for (int i = 0; i < NumberOfShapes; i++)
+	{
+		std::cout << "Shape " << i + 1 << std::endl;
+		std::cout << "Area: " << Array[i]->GetArea() << std::endl;
+		std::cout << "Perimeter: " << Array[i]->GetPerimeter() << std::endl;
+	}
+	RequestEnterFromInput();
+}
+
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::DeleteArray(FShape** Array, const int NumberOfShapes) const
+{
+	for (int i = 0; i < NumberOfShapes; i++)
+	{
+		delete Array[i];
+		Array[i] = nullptr;
+	}
+	delete Array;
+	Array = nullptr;
+}
+
+
+//-------------------------------------------------------------------------------------------
+//										 STATIC
+//-------------------------------------------------------------------------------------------
+void FShapeBuilder::AskForShapesStatic()
+{
+	system("CLS");
+	std::cout << "You will now be asked to provide information for 5 Shapes." << std::endl;
+
+	//Fill Array
+	TStaticArray<FShape*, 5> Shapes = FillStaticArray();
+
+	//Print Array
+	PrintArray(Shapes);
+	RequestEnterFromInput();
+
+	//Delete Array
+	DeleteArray(Shapes);
+
+	ActiveCommand = EMenuCommand::MainMenu;
+}
+
+//-------------------------------------------------------------------------------------------
+TStaticArray<FShape*, 5> FShapeBuilder::FillStaticArray() const
```

Comment:
This is fine and working perfectly. However, when we want to return something like a container, it may be better to use out parameters.
```
void FillStaticArray(TStaticArray<FShape*, 5>& OutArray) const;
```
because when you're creating an object for the return type and then, when you return, calling a copy assignment or copy constructor to assign the returned variable to your actual variable outside the function.

By using out parameters, you remove one of those copy operations.

Compilers are pretty smart these days and may choose to optimize this. However, you can't know for sure, so it may be better to not assume that the compiler will optimize the return type and use an out parameter instead.

We can discuss this in person if maybe the explanation is not too clear :)

Reply:
Yes, I saw that some extra copies were created, but at first I tried with references, though, being an object created in the method itself, that didn't work as planned. I forgot about OutParameters and then just defaulted to returning a copy instead 😢